pipeline.go

```diff
@@ -406,133 +443,106 @@ func (p Pipeline) BuildImages(force bool) error {
+	return p.runCommand(runConfig{
+		usePreconditions: false,
+		selection: func(step Step) bool {
+			return step.IsBuildable()
+		},
+		pre: func(runner Runner, step Step) error {
+			return nil
+		},
+		run: func(runner Runner, step Step) func() error {
+			return runner.ImageBuilder(step, force)
+		},
+		post: func(runner Runner, step Step) error {
+			return nil
+		},
+	})
 }
 
 // PullImages pulls all pullable images of Pipeline p in parallel.
 func (p Pipeline) PullImages(force bool) error {
-	pipelines, err := p.Definition.Pipelines()
-	if err != nil {
-		return err
-	}
-	abort := make(chan struct{})
-	runChannel := make(chan struct{})
-	var wg sync.WaitGroup
-	images := 0
-	durations := &sync.Map{}
-
-	for _, step := range pipelines.AllSteps() {
-		if step.IsBuildable() {
-			continue
-		}
-		wg.Add(1)
-		go runParallelPullImage(p.GetRunnerForMeta(step.Meta), step, force, durations, &wg, runChannel, abort)
-		images++
-	}
-
-	if Verbose {
-		pipelineLogger.Printf("Pull Images:")
-	}
-	start := time.Now()
-	close(runChannel)
-	wg.Wait()
-
-	if Verbose {
-		pipelineLogger.Printf("Pulled %d images in %s", images, time.Since(start))
-	}
-	var totalElapsedTime time.Duration
-	durations.Range(func(key, value interface{}) bool {
-		duration, ok := value.(time.Duration)
-		if ok {
-			totalElapsedTime += duration
-		}
-		return ok
+	return p.runCommand(runConfig{
+		usePreconditions: false,
+		selection: func(step Step) bool {
+			return step.IsPullable()
+		},
+		pre: func(runner Runner, step Step) error {
+			return nil
+		},
+		run: func(runner Runner, step Step) func() error {
+			return func() error {
+				if err := runner.ImageExistenceChecker(step)(); err != nil || force {
+					return runner.ImagePuller(step)()
+				}
+				return nil
+			}
+		},
+		post: func(runner Runner, step Step) error {
+			return nil
+		},
 	})
-	if Verbose {
-		pipelineLogger.Printf("Total time spent pulling images: %s", totalElapsedTime)
-	}
-	select {
-	case <-abort:
-		return fmt.Errorf("error while preparing images (pull)!")
-	default:
-	}
-	return nil
 }
 
 // KillContainers kills all running containers of Pipeline p.
 func (p Pipeline) KillContainers(preRun bool) error {
-	pipelines, err := p.Definition.Pipelines()
-	if err != nil {
-		return err
-	}
-	for _, pipeline := range *pipelines {
-		for _, step := range pipeline {
-			if preRun && step.Meta.KeepAlive == KeepAliveReplace {
-				continue
-			}
-			runner := p.GetRunnerForMeta(step.Meta)
-			if _, err := runner.ContainerKiller(step)(); err != nil {
-				pipelineLogger.Printf("Error killing %s: %s", step.ColoredName(), err)
-			}
-			if err := runner.ContainerRemover(step)(); err != nil {
-				pipelineLogger.Printf("Error removing %s: %s", step.ColoredName(), err)
+	return p.runCommand(runConfig{
+		usePreconditions: false,
+		selection: func(step Step) bool {
+			return !preRun || step.Meta.KeepAlive != KeepAliveReplace
+		},
+		pre: func(runner Runner, step Step) error {
+			return nil
```

Comment:
As above I feel like it would be more intuitive for `pre` itself to be `nil`

Reply:
Will be changed and simplified.